### PR TITLE
Add user editable key maps

### PR DESCRIPTION
**Pull request by franklin654**
* changed the Cmakelist to remove unnecessary space.
* added the functions to load and save key maps.
* created and installed event filters to pick up user input for custom key maps

### CODE REVIEW

**Comment by franklin654** [comment]
https://github.com/kitswas/VirtualGamePad-PC/assets/77013274/ea776fb9-8e35-4ff9-82ed-439bbd14d230


https://github.com/kitswas/VirtualGamePad-PC/assets/77013274/78ae1777-070b-47f8-a366-430692e6d240

working of the new user defined preferences feature.

**Comment by franklin654** [comment]
The interface

File: src/settings.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+{
+	QVariant value = settings->value(key);
+	return value;
+}
+
+void set_mouse_sensivity(int sensivity) {
+    mouse_sensivity = sensivity;
+}
+
+
```
Comment (kitswas) [inline src/settings.cpp:30]: This function is not used anywhere. Let's get rid of it.

```suggestion
```

**Comment by kitswas** [comment]
There's not a single line of documentation accompanying the code changes.  
This pull request won't be merged unless the new code is documented.

File: src/preferences.cpp [hunk 2]
```diff
@@ -0,0 +1,533 @@
...
+    ui->formLayout->setHorizontalSpacing(50);
+    ui->formLayout->setVerticalSpacing(10);
+	ui->horizontalSlider->setValue(mouse_sensivity / 100);
+    Preferences::load_keys();
+}
+
+void Preferences::set_button_icons()
+{
+    QPixmap x;
+    x.load("C:\\Users\\saisi\\Memory_card\\VirtualGamePad-PC\\res\\logos\\icons8-xbox-x-30.png");           // https://icons8.com/icon/80114/xbox-x
```
[the remark this answers is not in the document]
Comment (kitswas) [reply src/preferences.cpp]: Done.

**Comment by kitswas** [comment]
@franklin654 
Not a fan of commit 9a29227d9b90c4a416e8f70bc80484d061625a6a.

Before:

UI before commit

After:

UI after commit

Everything seems to be squished.  
You can decrease the default window size if the old UI looks too sparse to you. But revert that commit first.

Also, kindly [read what the QSizePolicies mean](https://doc.qt.io/qt-6/qsizepolicy.html#Policy-enum).

**Comment by kitswas** [comment]
And a bit of clarification about the documentation comments.

Something like

```
/**
 * @brief load_setting
 * get the value of a setting.
 * @param key
 * the fullname including the groups of the setting in string format
 * @return
 * returns the value as QVariant can be converted into almost all data types.
 */
QVariant load_setting(QString key)
```

ends up as:

image

So:

1. The function name should not be repeated after @brief.
2. Every sentence should begin with a Capital Letter.
3. Returns should not be written after @return.

Here's what I would write for this function:

```
/**
 * Get the value of a setting.
 * @param key
 * The full name including the groups of the setting as a string.
 * @return
 * The value as QVariant. It can be converted into almost all data types.
 */
QVariant load_setting(QString key)
```

image

See. Looks nicer already.

**Review by kitswas — APPROVED** [review]

**Comment by franklin654** [comment]
image
preview of the new interface

**Comment by kitswas** [comment]
Currently, there is no option to map a button to mouse input.  
We need to make it happen.

**Comment by kitswas** [comment]
We should support:
1. Left click
2. Right click
3. Mouse move

Here's what we could have in place of the `QLineEdit`s we are currently using:

image

The current `QLineEdit` could be moved to the keyboard tab.

image

**Comment by kitswas** [comment]
The Analogue Thumbsticks are missing.

For each thumbstick, we should be able to  
1. Map a key/mouse click to each direction, or
2. Map the whole thumbstick for mouse movement.

File: src/preferences.cpp [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+ * checks which object is sending the event and type of event.
+ * if event is a keyboard or mousebutton press than map and the object is buttonmap than get the virtual key code of the key pressed and store the change in a temporary variable.
+ * @param sender
+ * to get the address of the object that is triggering the event.
+ * @param event
+ * to capture the event that was triggered.
+ * @return [bool]
+ * true if event is handled else false.
+ */
+bool Preferences::eventFilter(QObject *sender, QEvent *event)
```
Comment (kitswas) [inline src/preferences.cpp:168]: Why is there so much repeated code?

File: src/preferences.cpp [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+                get_scan_code(VK_MBUTTON, buffer, 256);
+                break;
+            default:
+                qDebug() << "Some Error Occured No Legal Mouse Button found";
+                valid = false;
+            }
+            if(valid)
+            ui->dleftmap->setText(QString(buffer));
+        }
+    }
```
Comment (kitswas) [inline src/preferences.cpp:509]: Much of this code is simply duplicated with different variables.

This must change.

Comment (kitswas) [reply src/preferences.cpp:509]: I recommend creating a new component for input (like [the one discussed](https://github.com/kitswas/VirtualGamePad-PC/pull/5#issuecomment-1703814023)) and moving this logic there.

**Comment by kitswas** [comment]
@franklin654 Are you still working on this?

**Comment by franklin654** [comment]
> @franklin654 Are you still working on this?

Yes Just started today again.

File: src/preferences.cpp [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+{
+    char sc = MapVirtualKeyA((UINT)vk, MAPVK_VK_TO_CHAR);
+    if(sc >= '0' && sc <= 'Z') {
+        strncpy(a, "", size);
+        strncpy(a, &sc, 1);
+    }
+    else {
+        strncpy(a, vk_maps[vk], size);
+    }
+}
```
Comment (kitswas) [inline src/preferences.cpp:110]: @franklin654
There was a bug in the code using `strncpy` that was causing the application to crash.  
Did you fix that?  
Please debug and verify.

See `Preferences::get_scan_code()` in the file src/preferences.cpp.

Comment (franklin654) [reply src/preferences.cpp:110]: Yes I replaced the strncpy with strncpy_s with buffer size checks just haven't commited yet there are other changes that need to be done in preferences.cpp

**Comment by kitswas** [comment]
@franklin654
Please merge changes from main into your branch and run the code linter.
```bash
./code-lint.ps1
```

**Pull request merged**